Improve test suite interface.

marler8997 opened the pull request:
Makes the test suite a little more accommodating to developers who don't know how to use it. It now indicates that the suite is meant to be run through the Makefile (not by running d_do_test.d directly) and also prints a nice error message when environment variables are missing that would have been set by the Makefile.  This is a nice contrast to what was done before which would have been to fail at some later point in the test with a cryptic error message as a result of some environment variable being empty.

wilzbach commented on test/d_do_test.d:
```diff
@@ -436,7 +440,25 @@ bool compareOutput(string output, string refoutput)
     }
 }
 
+string envGetRequired(in char[] name)
```
why not `string`?

marler8997 replied:
`const char[]` conveys that the contents of the string do not need to be preserved. If you look at `environment.get` you'll see it also uses `in char[]`.  A function that accepts `string` implies that it requires the contents of that string to remain unchanged for the life of the program, however, the `environment.get` function, and this function do not have this requirement.

wilzbach replied:
Well in this case you only call this function with pre-defined strings which are immutable by definition ;-)

wilzbach commented on test/d_do_test.d:
```diff
@@ -27,15 +30,16 @@ void usage()
-          ~ "      EXE:          .exe or <null>\n");
+          ~ "      SEP:           \\ or / (required)\n"
+          ~ "      OBJ:          .obj or .o (required)\n"
+          ~ "      EXE:          .exe or <null> (required)\n");
```
we could auto-detect the default value here.
For example:

```
version(windows) string exe = ".exe";
else string exe = "";
```

marler8997 replied:
Setting reasonable defaults was actually my first thought.  The thinking was it might be nice to make it easier to call `d_do_test.d` on it's own without the Makefile.  However, I realized that `d_do_test.d` is not designed to run on it's own, it's really just an extension for the Makefile, and the reason it's using environment variables in the first place is because the Makefile is not just using those variables to call d_do_test, but also for other parts of it's configuration.

If we wanted to allow the developer to call `d_do_test.d` on it's own I would say let's put defaults in there, but if it's just an extension for the Makefile then I think it makes more sense to just verify the interface between the Makefile and d_do_test (make sure required variables are given).

wilzbach commented on test/d_do_test.d:
```diff
@@ -27,15 +30,16 @@ void usage()
           ~ "      ARGS:          set to execute all combinations of\n"
           ~ "      REQUIRED_ARGS: arguments always passed to the compiler\n"
-          ~ "      DMD:           compiler to use, ex: ../src/dmd\n"
+          ~ "      DMD:           compiler to use, ex: ../src/dmd (required)\n"
```
`../src/dmd` no longer exists ;-)

marler8997 replied:
Yes I believe you just recently changed that because of the `ddmd` to `dmd` change :)  A nice change by the way.

wilzbach reviewed (commented):
Why not setting reasonable defaults?

wilzbach commented on test/d_do_test.d:
```diff
@@ -17,6 +17,9 @@ import core.sys.posix.sys.wait;
     write("d_do_test <input_dir> <test_name> <test_extension>\n"
+          ~ "\n"
+          ~ "   Note: this program is normally called through the Makefile, it"
+          ~ "         is not meant to be called directly by the user\n"
```
Nitpick: end sentences with a dot.

wilzbach commented on test/d_do_test.d:
```diff
@@ -436,7 +440,25 @@ bool compareOutput(string output, string refoutput)
+    return value;
+}
+
+class SilentQuit : Exception { this() { super(null); } }
```
nit: new line

wilzbach commented on test/d_do_test.d:
```diff
@@ -436,7 +440,25 @@ bool compareOutput(string output, string refoutput)
+{
+    try { return tryMain(args); }
+    catch(SilentQuit) { return 1; }
+}
```
nit: newline

wilzbach reviewed (approved):
A couple of nits, but I think this is quite helpful as getting the DMD testsuite to run is really painful for newcomers...

marler8997 committed:
Improve test suite interface.

Makes the test suite a little more accomodating to developers who don't know how to use it yet. It now indicatest that the suite is meant to be run through the makefile (not by running d_do_test.d directory) and also prints a nice error message when environment variables are missing that would have been set by the Makefile.  This is a nice contrast to what was done before which would have been to fail at some later point in the test with a cryptic error message as a result of some environment variable being empty.
